### Код из лекции
### Первая часть

In [1]:
from keras.datasets import imdb

(train_data, train_labels), (test_data, test_labels) = imdb.load_data(
    num_words=10000)

D:\Anaconda\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


In [2]:
import numpy as np

def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        results[i, sequence] = 1.
    return results

x_train = vectorize_sequences(train_data)
x_test = vectorize_sequences(test_data)


In [3]:
y_train = np.asarray(train_labels).astype('float32')
y_test = np.asarray(test_labels).astype('float32')

In [4]:
from keras import models
from keras import layers

model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

In [5]:
from keras import optimizers
from keras import losses
from keras import metrics

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

In [6]:
x_val = x_train[:5000]
partial_x_train = x_train[5000:]
y_val = y_train[:5000]
partial_y_train = y_train[5000:]
print(partial_x_train.shape)
print(x_val.shape)

(20000, 10000)
(5000, 10000)


In [7]:
history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=4,
                    batch_size=512,
                    validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Train on 20000 samples, validate on 5000 samples
Epoch 1/4
20000/20000 [==============================] - 4s 178us/step - loss: 0.4715 - acc: 0.8056 - val_loss: 0.3695 - val_acc: 0.8568
Epoch 2/4
20000/20000 [==============================] - 3s 127us/step - loss: 0.2722 - acc: 0.9049 - val_loss: 0.2942 - val_acc: 0.8862
Epoch 3/4
20000/20000 [==============================] - 3s 125us/step - loss: 0.2027 - acc: 0.9307 - val_loss: 0.2700 - val_acc: 0.8910
Epoch 4/4
25000/25000 [==============================] - 4s 153us/step


In [8]:
print(results)
print(np.array(history.history['val_acc']).mean())

[0.30405114110946657, 0.87916]
0.8809999998807907


-You used two hidden layers. Try using one or three hidden layers, and see how doing so affects validation and test accuracy.
### с тремя скрытыми слоями

In [9]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Train on 20000 samples, validate on 5000 samples
Epoch 1/4
20000/20000 [==============================] - 3s 139us/step - loss: 0.5295 - acc: 0.7844 - val_loss: 0.3934 - val_acc: 0.8542
Epoch 2/4
20000/20000 [==============================] - 2s 122us/step - loss: 0.2923 - acc: 0.9042 - val_loss: 0.2911 - val_acc: 0.8910
Epoch 3/4
20000/20000 [==============================] - 2s 124us/step - loss: 0.2108 - acc: 0.9274 - val_loss: 0.2938 - val_acc: 0.8830
Epoch 4/4
25000/25000 [==============================] - 4s 165us/step


In [10]:
print(results)
print(np.array(history.history['val_acc']).mean())

[0.31313087109565735, 0.87616]
0.879200000166893


# с одним скрытым слоем

In [11]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Train on 20000 samples, validate on 5000 samples
Epoch 1/4
20000/20000 [==============================] - 3s 168us/step - loss: 0.4802 - acc: 0.8140 - val_loss: 0.3691 - val_acc: 0.8764
Epoch 2/4
20000/20000 [==============================] - 3s 125us/step - loss: 0.3024 - acc: 0.9043 - val_loss: 0.3024 - val_acc: 0.8856
Epoch 3/4
20000/20000 [==============================] - 3s 126us/step - loss: 0.2339 - acc: 0.9246 - val_loss: 0.2799 - val_acc: 0.8916
Epoch 4/4
25000/25000 [==============================] - 4s 140us/step


In [12]:
print(results)
print(np.array(history.history['val_acc']).mean())

[0.3138272501039505, 0.87164]
0.8841499995231629


Результаты следуюшие:

при одном скрытом слое: test_acc = 0.87164, val_acc = 0.8841499995231629

при двух скрытых слоях: test_acc = 0.87916, val_acc = 0.8809999998807907

при трех скрытый слоях: test_acc = 0.87616, val_acc = 0.879200000166893

Лучший показатель test accuracy - при 2 слоях, validation accuracy - при 1 слое. 

____________________________________
-Try using layers with more hidden units or fewer hidden units: 32 units, 64 units, and so on.

In [13]:
model = models.Sequential()
model.add(layers.Dense(32, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Train on 20000 samples, validate on 5000 samples
Epoch 1/4
20000/20000 [==============================] - 3s 157us/step - loss: 0.4525 - acc: 0.8065 - val_loss: 0.3561 - val_acc: 0.8542
Epoch 2/4
20000/20000 [==============================] - 3s 131us/step - loss: 0.2527 - acc: 0.9090 - val_loss: 0.3636 - val_acc: 0.8450
Epoch 3/4
20000/20000 [==============================] - 3s 130us/step - loss: 0.1926 - acc: 0.9293 - val_loss: 0.2845 - val_acc: 0.8880
Epoch 4/4
25000/25000 [==============================] - 4s 158us/step


In [14]:
print(results)
print(np.array(history.history['val_acc']).mean())

[0.3310349017906189, 0.87428]
0.8684000006198884


In [15]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)
print(results)
print(np.array(history.history['val_acc']).mean())

Train on 20000 samples, validate on 5000 samples
Epoch 1/4
20000/20000 [==============================] - 4s 178us/step - loss: 0.4560 - acc: 0.7893 - val_loss: 0.3105 - val_acc: 0.8842
Epoch 2/4
20000/20000 [==============================] - 3s 155us/step - loss: 0.2435 - acc: 0.9085 - val_loss: 0.3807 - val_acc: 0.8426
Epoch 3/4
20000/20000 [==============================] - 3s 159us/step - loss: 0.1803 - acc: 0.9331 - val_loss: 0.3366 - val_acc: 0.8730
Epoch 4/4
25000/25000 [==============================] - 5s 197us/step
[0.3138675223207474, 0.87924]
0.8729500000238418


In [16]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)
print(results)
print(np.array(history.history['val_acc']).mean())

Train on 20000 samples, validate on 5000 samples
Epoch 1/4
20000/20000 [==============================] - 4s 208us/step - loss: 0.4473 - acc: 0.7998 - val_loss: 0.3019 - val_acc: 0.8860
Epoch 2/4
20000/20000 [==============================] - 4s 195us/step - loss: 0.2463 - acc: 0.9086 - val_loss: 0.3054 - val_acc: 0.8756
Epoch 3/4
20000/20000 [==============================] - 4s 191us/step - loss: 0.1847 - acc: 0.9318 - val_loss: 0.3056 - val_acc: 0.8794
Epoch 4/4
25000/25000 [==============================] - 5s 197us/step
[0.3236232180309296, 0.87628]
0.8822999999046326


In [17]:
model = models.Sequential()
model.add(layers.Dense(8, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(8, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)
print(results)
print(np.array(history.history['val_acc']).mean())

Train on 20000 samples, validate on 5000 samples
Epoch 1/4
20000/20000 [==============================] - 3s 148us/step - loss: 0.5089 - acc: 0.7840 - val_loss: 0.3781 - val_acc: 0.8660
Epoch 2/4
20000/20000 [==============================] - 2s 121us/step - loss: 0.3001 - acc: 0.8990 - val_loss: 0.3250 - val_acc: 0.8680
Epoch 3/4
20000/20000 [==============================] - 2s 120us/step - loss: 0.2270 - acc: 0.9259 - val_loss: 0.2770 - val_acc: 0.8932
Epoch 4/4
25000/25000 [==============================] - 3s 128us/step
[0.29319496866226197, 0.88296]
0.8790500003099442


In [18]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(8, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)
print(results)
print(np.array(history.history['val_acc']).mean())

Train on 20000 samples, validate on 5000 samples
Epoch 1/4
20000/20000 [==============================] - 3s 156us/step - loss: 0.4688 - acc: 0.8054 - val_loss: 0.3648 - val_acc: 0.8624
Epoch 2/4
20000/20000 [==============================] - 2s 124us/step - loss: 0.2775 - acc: 0.9087 - val_loss: 0.2952 - val_acc: 0.8844
Epoch 3/4
20000/20000 [==============================] - 2s 122us/step - loss: 0.2089 - acc: 0.9304 - val_loss: 0.2688 - val_acc: 0.8930
Epoch 4/4
25000/25000 [==============================] - 3s 139us/step
[0.4351994038391113, 0.82552]
0.8698499992370605


In [19]:
model = models.Sequential()
model.add(layers.Dense(8, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)
print(results)
print(np.array(history.history['val_acc']).mean())

Train on 20000 samples, validate on 5000 samples
Epoch 1/4
20000/20000 [==============================] - 3s 150us/step - loss: 0.4879 - acc: 0.8084 - val_loss: 0.3766 - val_acc: 0.8718
Epoch 2/4
20000/20000 [==============================] - 2s 120us/step - loss: 0.3063 - acc: 0.9033 - val_loss: 0.3082 - val_acc: 0.8830
Epoch 3/4
20000/20000 [==============================] - 2s 120us/step - loss: 0.2403 - acc: 0.9244 - val_loss: 0.2870 - val_acc: 0.8878
Epoch 4/4
25000/25000 [==============================] - 3s 127us/step
[0.29175043167114256, 0.88292]
0.8835999998092651


Вывод по hidden units для данных данных:
- чем их меньше, тем лучше для test accuracy
- при двух слоях модель наилучшие показатели test accuracy = 0.88296 при 8 нейронах
- но test accuracy еще лучше, если в модели один скрытый слой с 8 нейронами (0,88292)
- validation accuracy наилучший при 1 слое и 8 нейронах (0.8835999998092651)

________________________________
-Try using the mse loss function instead of binary_crossentropy.

In [20]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='mean_squared_error',
              metrics=['accuracy'])

history = model.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Train on 20000 samples, validate on 5000 samples
Epoch 1/4
20000/20000 [==============================] - 3s 155us/step - loss: 0.1504 - acc: 0.8098 - val_loss: 0.1108 - val_acc: 0.8598
Epoch 2/4
20000/20000 [==============================] - 2s 122us/step - loss: 0.0792 - acc: 0.9098 - val_loss: 0.0911 - val_acc: 0.8790
Epoch 3/4
20000/20000 [==============================] - 2s 123us/step - loss: 0.0586 - acc: 0.9339 - val_loss: 0.0860 - val_acc: 0.8840
Epoch 4/4
25000/25000 [==============================] - 4s 144us/step


In [21]:
print(results)
print(np.array(history.history['val_acc']).mean())

[0.09008973398208618, 0.87684]
0.877850000333786


Вывод: 
- при binary_crossentropy
- test accuracy = 0.87916, val_acc = 0.8809999998807907
- при mse
- test accuracy = 0.87684, val_acc = 0.877850000333786
- лучше при binary_crossentropy, но разница в тысячных долях
______________________
-Try using the tanh activation (an activation that was popular in the early days of neural networks) instead of relu.

In [22]:
model = models.Sequential()
model.add(layers.Dense(16, activation='tanh', input_shape=(10000,)))
model.add(layers.Dense(16, activation='tanh'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)
print(results)
print(np.array(history.history['val_acc']).mean())

Train on 20000 samples, validate on 5000 samples
Epoch 1/4
20000/20000 [==============================] - 3s 159us/step - loss: 0.4698 - acc: 0.8165 - val_loss: 0.3422 - val_acc: 0.8796
Epoch 2/4
20000/20000 [==============================] - 2s 124us/step - loss: 0.2656 - acc: 0.9106 - val_loss: 0.2959 - val_acc: 0.8818
Epoch 3/4
20000/20000 [==============================] - 2s 124us/step - loss: 0.1922 - acc: 0.9324 - val_loss: 0.2859 - val_acc: 0.8838
Epoch 4/4
25000/25000 [==============================] - 3s 139us/step
[0.3243111707830429, 0.87588]
0.8823999998092651


- relu: test accuracy = 0.87916, val_acc = 0.8809999998807907
- tanh: test accuracy = 0.87588, val_acc = 0.8823999998092651
при relu test accuracy лучше, а val_acc хуже, но разница опять в тысячных долях 

### Вторая часть 

### Код из лекции

In [23]:
from keras.datasets import reuters

(train_data, train_labels), (test_data, test_labels) = reuters.load_data(
    num_words=10000)

In [24]:
import numpy as np

def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        results[i, sequence] = 1.
    return results

x_train = vectorize_sequences(train_data)
x_test = vectorize_sequences(test_data)


In [25]:
def to_one_hot(labels, dimension=46):
    results = np.zeros((len(labels), dimension))
    for i, label in enumerate(labels):
        results[i, label] = 1.
    return results

one_hot_train_labels = to_one_hot(train_labels)
one_hot_test_labels = to_one_hot(test_labels)


In [26]:
from keras.utils.np_utils import to_categorical

one_hot_train_labels = to_categorical(train_labels)
one_hot_test_labels = to_categorical(test_labels)

In [27]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

x_val = x_train[:1000]
partial_x_train = x_train[1000:]

y_val = one_hot_train_labels[:1000]
partial_y_train = one_hot_train_labels[1000:]

history = model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
print(results)
print(np.array(history.history['val_acc']).mean())

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 3s 328us/step - loss: 2.5322 - acc: 0.4955 - val_loss: 1.7208 - val_acc: 0.6120
Epoch 2/9
7982/7982 [==============================] - 2s 191us/step - loss: 1.4452 - acc: 0.6879 - val_loss: 1.3459 - val_acc: 0.7060
Epoch 3/9
7982/7982 [==============================] - 2s 190us/step - loss: 1.0953 - acc: 0.7651 - val_loss: 1.1708 - val_acc: 0.7430
Epoch 4/9
7982/7982 [==============================] - 1s 182us/step - loss: 0.8697 - acc: 0.8165 - val_loss: 1.0793 - val_acc: 0.7590
Epoch 5/9
7982/7982 [==============================] - 1s 154us/step - loss: 0.7034 - acc: 0.8472 - val_loss: 0.9844 - val_acc: 0.7810
Epoch 6/9
7982/7982 [==============================] - 1s 144us/step - loss: 0.5667 - acc: 0.8802 - val_loss: 0.9411 - val_acc: 0.8040
Epoch 7/9
7982/7982 [==============================] - 1s 150us/step - loss: 0.4581 - acc: 0.9048 - val_loss: 0.9083 - val_acc: 0.8020
Epoch 8

- Try using larger or smaller layers: 32 units, 128 units, and so on.

In [28]:
model = models.Sequential()
model.add(layers.Dense(32, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
print(results)
print(np.array(history.history['val_acc']).mean())

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 2s 225us/step - loss: 3.0975 - acc: 0.4546 - val_loss: 2.3165 - val_acc: 0.6090
Epoch 2/9
7982/7982 [==============================] - 1s 125us/step - loss: 1.8966 - acc: 0.6625 - val_loss: 1.6682 - val_acc: 0.6590
Epoch 3/9
7982/7982 [==============================] - 1s 126us/step - loss: 1.4035 - acc: 0.7175 - val_loss: 1.4031 - val_acc: 0.6950
Epoch 4/9
7982/7982 [==============================] - 1s 126us/step - loss: 1.1576 - acc: 0.7502 - val_loss: 1.2633 - val_acc: 0.7280
Epoch 5/9
7982/7982 [==============================] - 1s 129us/step - loss: 0.9929 - acc: 0.7890 - val_loss: 1.1801 - val_acc: 0.7520
Epoch 6/9
7982/7982 [==============================] - 1s 124us/step - loss: 0.8635 - acc: 0.8166 - val_loss: 1.1190 - val_acc: 0.7720
Epoch 7/9
7982/7982 [==============================] - 1s 126us/step - loss: 0.7562 - acc: 0.8356 - val_loss: 1.0802 - val_acc: 0.7700
Epoch 8

In [29]:
model = models.Sequential()
model.add(layers.Dense(128, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
print(results)
print(np.array(history.history['val_acc']).mean())

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 3s 332us/step - loss: 2.2531 - acc: 0.5534 - val_loss: 1.4524 - val_acc: 0.6630
Epoch 2/9
7982/7982 [==============================] - 2s 237us/step - loss: 1.1717 - acc: 0.7428 - val_loss: 1.1415 - val_acc: 0.7490
Epoch 3/9
7982/7982 [==============================] - 2s 234us/step - loss: 0.8202 - acc: 0.8260 - val_loss: 1.0116 - val_acc: 0.7760
Epoch 4/9
7982/7982 [==============================] - 2s 238us/step - loss: 0.5955 - acc: 0.8776 - val_loss: 0.9961 - val_acc: 0.7720
Epoch 5/9
7982/7982 [==============================] - 2s 232us/step - loss: 0.4333 - acc: 0.9114 - val_loss: 0.9036 - val_acc: 0.7960
Epoch 6/9
7982/7982 [==============================] - 2s 242us/step - loss: 0.3378 - acc: 0.9263 - val_loss: 0.8720 - val_acc: 0.8190
Epoch 7/9
7982/7982 [==============================] - 2s 234us/step - loss: 0.2604 - acc: 0.9411 - val_loss: 0.8659 - val_acc: 0.8270
Epoch 8

In [30]:
model = models.Sequential()
model.add(layers.Dense(256, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
print(results)
print(np.array(history.history['val_acc']).mean())

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 4s 439us/step - loss: 2.0758 - acc: 0.5719 - val_loss: 1.2636 - val_acc: 0.7210
Epoch 2/9
7982/7982 [==============================] - 2s 276us/step - loss: 0.9763 - acc: 0.7873 - val_loss: 1.0148 - val_acc: 0.7710
Epoch 3/9
7982/7982 [==============================] - 2s 311us/step - loss: 0.6277 - acc: 0.8653 - val_loss: 0.8761 - val_acc: 0.8160
Epoch 4/9
7982/7982 [==============================] - 3s 341us/step - loss: 0.4272 - acc: 0.9097 - val_loss: 0.8602 - val_acc: 0.8100
Epoch 5/9
7982/7982 [==============================] - 3s 342us/step - loss: 0.3064 - acc: 0.9332 - val_loss: 0.8797 - val_acc: 0.8190
Epoch 6/9
7982/7982 [==============================] - 3s 347us/step - loss: 0.2276 - acc: 0.9476 - val_loss: 0.8883 - val_acc: 0.8220
Epoch 7/9
7982/7982 [==============================] - 3s 348us/step - loss: 0.1995 - acc: 0.9470 - val_loss: 0.8960 - val_acc: 0.8240
Epoch 8

In [31]:
model = models.Sequential()
model.add(layers.Dense(128, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
print(results)
print(np.array(history.history['val_acc']).mean())

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 3s 385us/step - loss: 2.2850 - acc: 0.5593 - val_loss: 1.4532 - val_acc: 0.6740
Epoch 2/9
7982/7982 [==============================] - 2s 232us/step - loss: 1.1340 - acc: 0.7521 - val_loss: 1.1224 - val_acc: 0.7580
Epoch 3/9
7982/7982 [==============================] - 2s 209us/step - loss: 0.7916 - acc: 0.8314 - val_loss: 0.9898 - val_acc: 0.7870
Epoch 4/9
7982/7982 [==============================] - 2s 221us/step - loss: 0.5786 - acc: 0.8822 - val_loss: 0.9744 - val_acc: 0.7890
Epoch 5/9
7982/7982 [==============================] - 2s 233us/step - loss: 0.4247 - acc: 0.9118 - val_loss: 0.9210 - val_acc: 0.7990
Epoch 6/9
7982/7982 [==============================] - 2s 242us/step - loss: 0.3330 - acc: 0.9316 - val_loss: 0.8707 - val_acc: 0.8210
Epoch 7/9
7982/7982 [==============================] - 2s 235us/step - loss: 0.2629 - acc: 0.9401 - val_loss: 0.8734 - val_acc: 0.8260
Epoch 8

- 32 нейрона: test_acc = 0.7622439893143366, val_acc = 0.7176666611565484
- 64 нейрона: test_acc = 0.7880676759212865, val_acc = 0.756111107243432
- 128 нейронов: test_acc = 0.7684772929652716, val_acc = 0.7772222203678555
- 128(1) - 64(2) нейрона: test_acc = 0.7898486197950154, val_acc = 0.7880000003708734
- 256(1) - 128(2) нейронов: test_acc = 0.800979519145147, val_acc = 0.8018888918558756 - лучшие показатели

- You used two hidden layers. Now try using a single hidden layer, or three hidden layers.

In [32]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
print(results)
print(np.array(history.history['val_acc']).mean())

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 2s 279us/step - loss: 2.6321 - acc: 0.5487 - val_loss: 1.8462 - val_acc: 0.6360
Epoch 2/9
7982/7982 [==============================] - 1s 180us/step - loss: 1.5211 - acc: 0.7127 - val_loss: 1.3738 - val_acc: 0.7100
Epoch 3/9
7982/7982 [==============================] - 1s 182us/step - loss: 1.1090 - acc: 0.7772 - val_loss: 1.1511 - val_acc: 0.7450
Epoch 4/9
7982/7982 [==============================] - 1s 183us/step - loss: 0.8635 - acc: 0.8256 - val_loss: 1.0251 - val_acc: 0.7860
Epoch 5/9
7982/7982 [==============================] - 1s 180us/step - loss: 0.6935 - acc: 0.8626 - val_loss: 0.9478 - val_acc: 0.8010
Epoch 6/9
7982/7982 [==============================] - 1s 182us/step - loss: 0.5650 - acc: 0.8918 - val_loss: 0.8866 - val_acc: 0.8230
Epoch 7/9
7982/7982 [==============================] - 1s 182us/step - loss: 0.4650 - acc: 0.9100 - val_loss: 0.8535 - val_acc: 0.8160
Epoch 8

In [33]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
print(results)
print(np.array(history.history['val_acc']).mean())

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 2s 264us/step - loss: 2.6430 - acc: 0.4728 - val_loss: 1.7158 - val_acc: 0.6230
Epoch 2/9
7982/7982 [==============================] - 1s 147us/step - loss: 1.4251 - acc: 0.6850 - val_loss: 1.3346 - val_acc: 0.7160
Epoch 3/9
7982/7982 [==============================] - 1s 145us/step - loss: 1.0801 - acc: 0.7548 - val_loss: 1.1593 - val_acc: 0.7430
Epoch 4/9
7982/7982 [==============================] - 1s 145us/step - loss: 0.8684 - acc: 0.8043 - val_loss: 1.0689 - val_acc: 0.7640
Epoch 5/9
7982/7982 [==============================] - 1s 144us/step - loss: 0.6944 - acc: 0.8444 - val_loss: 1.0847 - val_acc: 0.7670
Epoch 6/9
7982/7982 [==============================] - 1s 144us/step - loss: 0.5572 - acc: 0.8753 - val_loss: 1.0184 - val_acc: 0.7890
Epoch 7/9
7982/7982 [==============================] - 1s 145us/step - loss: 0.4561 - acc: 0.8988 - val_loss: 0.9697 - val_acc: 0.8030
Epoch 8

 - 1 слой: test_acc = 0.8023152271234235, val_acc = 0.7737777768241035 - лучшие показатели
 - 2 слоя: test_acc = 0.7880676759212865, val_acc = 0.756111107243432
 - 3 слой: test_acc = 0.7756010685663401, val_acc = 0.7566666629579332

Общий вывод: лучшие показатели - это, конечно, хорошо, но переобучуться не хочется, тем более нагружать модель ради сотых или тысячных доль значений на несложных данных нет смысла. Поэтому в первой части ноутбука лучше остановиться на 2 слоях и 16 нейронках, а во второй части - на 2 слоях и 64 нейронках (при тех же показателях, которые были на лекции). 